### PR TITLE
Feature/optional chaining

### LINKED ISSUE #10227: Unexpected token error at code with optional chaining when using Typescript
<s>UPD for everyone who will find this issue. For fix this error acornjs/acorn#891 and acornjs/acorn#890 must be merged.
UPD2 we all awaiting while estree/estree/pull/204 will be merged</s>

UPD3 fix for optional chaining landed into acorn 7.3.0 but webpack@4 uses acorn 6.x. so fix of this issue will be in webpack@5: https://github.com/webpack/webpack/issues/10227#issuecomment-642734920 

### For Babel
Add `@babel/plugin-proposal-optional-chaining` and `@babel/plugin-proposal-nullish-coalescing-operator` plugins in your config

### For Typescript
Do not upgrade to version 3.8 or set compiler option `target` to `es2019`

### Yarn/NPM level
See the https://github.com/webpack/webpack/issues/10227#issuecomment-653408484

---

**Do you want to request a *feature* or report a *bug*?**
I don't know is it a bug or feature

**What is the current behavior?**

Get error "Unecpected token"

**What is the expected behavior?**

Build bundle successfull

**Description**
When I set in `tsconfig.json` `target: "ESNext"` TS keeps my code with optional chaining (and nullish coalescing)  as is. But I've got an error
```
Module parse failed: Unexpected token (2:10)
File was processed with these loaders:
 * ./node_modules/ts-loader/index.js
You may need an additional loader to handle the result of these loaders.
| var a = { b: 2 };
> var c = a?.b;
```

I made a minimum reproduce repo [inoyakaigor/webpack-typescript-optional-chaining](/inoyakaigor/webpack-typescript-optional-chaining)

How I can avoid this error?

### LINKED ISSUE #11186: [webpack 5] Identifier of ?. is not correctly parsed
# Bug report

**What is the current behavior?**

acorn's ChainExpression is not correctly parsed. Variables are not correctly resolved to imports when optional chaining is used.

```javascript
import test from './test';

console.log(test.a)
console.log(test?.a)
```

compiles to

```javascript
/* harmony import */ var _test__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./test */ "./src/test.js");


console.log(_test__WEBPACK_IMPORTED_MODULE_0__.default.a)
console.log(test?.a)
```

**If the current behavior is a bug, please provide the steps to reproduce.**

https://github.com/SunriseFox/webpack-bug-optional-chaining

`yarn && yarn start`

**What is the expected behavior?**

should be 

```
/* harmony import */ var _test__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./test */ "./src/test.js");


console.log(_test__WEBPACK_IMPORTED_MODULE_0__.default.a)
console.log(_test__WEBPACK_IMPORTED_MODULE_0__.default?.a)
```

**Other relevant information:**
webpack version:  5.0.0-beta.22
Node.js version: v12.16.3
Operating System: MacOS 10.15.5
Additional tools: none

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**

closes #11186 
closes #10227 

cherry-pick from  https://github.com/webpack/webpack/pull/11198

**Did you add tests for your changes?**

yes

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

nothing

### CODE REVIEW

File: test/cases/parsing/optional-chaining/test.filter.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+const supportsOptionalChaining = require("../../../helpers/supportsOptionalChaining");
+
+/**
+ * @param {import("../../../../").Configuration} config
+ * @returns {boolean}
+ */
+module.exports = function (config) {
+	if (config.mode === "production") return false;
```
Comment (vankop) [inline test/cases/parsing/optional-chaining/test.filter.js:8]: Terser does not supports optional chaining yet.
https://github.com/terser/terser/issues/567

nullish coalescing works on try.terser.org, but not optimized.

File: lib/ConstPlugin.js [hunk 2]
```diff
@@ -420,7 +420,10 @@ class ConstPlugin {
 								//
 								// ------------------------------------------
 								//
-								const dep = new ConstDependency(" undefined", expr.range);
+								const dep = new ConstDependency(
+									evaluated.isNull() ? " null" : " undefined",
```
Comment (vankop) [inline lib/ConstPlugin.js:424]: as I understand from babel https://babeljs.io/repl#?browsers=chrome%20%3E%2088&build=&builtIns=false&spec=true&loose=true&code_lz=MYewdgzgLgBAhjAvDA3gXwNwCgtwHQBGA_HnEA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ces2015%2Creact%2Cstage-2%2Ctypescript%2Cenv&prettier=false&targets=Node-12&version=7.11.1&externalPlugins=

undefined should be used

Comment (vankop) [reply lib/ConstPlugin.js:424]: btw `"void 0".length < "undefined".length`

Comment (sokra) [reply lib/ConstPlugin.js:424]: interesting... I'll fix that

**Comment by sokra** [comment]
Thanks

**Comment by dreyks** [comment]
Thanks!

Is there an estimation on when this will be released?

**Comment by chenxsan** [comment]
@dreyks It's already released https://github.com/webpack/webpack/releases/tag/v5.0.0-beta.25

**Comment by dreyks** [comment]
will it be ported to the stable branch or will it only be available in webpack 5?

**Comment by evilebottnawi** [comment]
@dreyks only for webpack@5

**Comment by dreyks** [comment]
btw terser [5.2.0](https://github.com/terser/terser/blob/e0c1a2b50921f549b56b3cdb3ea145a377c760e8/CHANGELOG.md#v520) is out now with the optional chaining support

**Pull request merged**